Foundation/URLCache.swift

```diff
@@ -243,11 +283,91 @@ open class URLCache : NSObject {
         usage of the on-disk cache. 
         @result the current usage of the on-disk cache of the receiver.
     */
-    open var currentDiskUsage: Int { NSUnimplemented() }
+    open var currentDiskUsage: Int {
+        return self.syncQ.sync { self.persistence?.cacheSizeInDesk ?? 0 }
+    }
+
 }
 
 extension URLCache {
     public func storeCachedResponse(_ cachedResponse: CachedURLResponse, for dataTask: URLSessionDataTask) { NSUnimplemented() }
     public func getCachedResponse(for dataTask: URLSessionDataTask, completionHandler: (CachedURLResponse?) -> Void) { NSUnimplemented() }
     public func removeCachedResponse(for dataTask: URLSessionDataTask) { NSUnimplemented() }
 }
+
+fileprivate struct _CachePersistence {
+
+    let path: String
+
+    var cacheSizeInDesk: Int {
+        do {
+            let subFiles = try FileManager.default.subpathsOfDirectory(atPath: path)
+            var total: Int = 0
+            for fileName in subFiles {
+                let attributes = try FileManager.default.attributesOfItem(atPath: path.appending(fileName))
+                total += (attributes[.size] as? Int) ?? 0
+            }
+            return total
+        } catch {
+            return 0
+        }
+    }
+
+    func setupCacheDirectory() {
+        do {
+            if FileManager.default.fileExists(atPath: path) {
+                try FileManager.default.removeItem(atPath: path)
+            }
+            
+            try FileManager.default.createDirectory(atPath: path, withIntermediateDirectories: true)
+        } catch {
+            fatalError("Unable to create directories for URLCache: \(error.localizedDescription)")
+        }
+    }
+
+    func saveCachedResponse(_ response: CachedURLResponse, for request: URLRequest) {
+        let archivedData = NSKeyedArchiver.archivedData(withRootObject: response)
+        do {
+            let cacheFileURL = urlForFileIdentifier(request.cacheFileIdentifier)
+            try archivedData.write(to: cacheFileURL)
+        } catch {
+            fatalError("Unable to save cache data: \(error.localizedDescription)")
+        }
+    }
+
+    func cachedResponse(for request: URLRequest) -> CachedURLResponse? {
+        let url = urlForFileIdentifier(request.cacheFileIdentifier)
+        guard let data = try? Data(contentsOf: url),
+            let response = NSKeyedUnarchiver.unarchiveObject(with: data) as? CachedURLResponse else {
+            return nil
+        }
+
+        return response
+    }
+
+    func deleteAllCaches() {
+        do {
+            try FileManager.default.removeItem(atPath: path)
+        } catch {
+            fatalError("Unable to flush database for URLCache: \(error.localizedDescription)")
+        }
+    }
+
+    private func urlForFileIdentifier(_ identifier: String) -> URL {
+        return URL(fileURLWithPath: path + "/" + identifier)
+    }
+
+}
+
+fileprivate extension URLRequest {
+
+    var cacheFileIdentifier: String {
```

Comment:
I have a slightly different design here (I use a `URLCacheKey` type, instead of `String`), but one important thing is that generating a key from the `URLRequest` depends on some details from the request beside the URL itself. So my equivalent method returns an optional.

The first thing is the protocol. Protocols like `data:` or `file:` should not be cached for obvious reasons. I also think unknown protocols should not be cached either.

Also, for known protocols like `http:` or `https:` (and maybe for `ftp:`, but I don’t know that protocol that well), we should do a couple more things. For example:
- I don’t think you can do a HTTP request without a method, so that's a `return nil` in the case the protocol is HTTP.
- I think some headers should be taken into account. The spec refers to `User-Agent`, but others like `Accept-Language` and `Range` might be interesting.
- I think the hash of the body contents should also be used.
- I also think one should not try to cache request with body streams, because that will consume it, which is not good.
- Finally I would hash everything and return an hexadecimal string, to have a valid file name.